File: rewrite-csharp/src/main/java/org/openrewrite/csharp/CsDocCommentVisitor.java
/*
 * Copyright 2024 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.csharp;

import org.jspecify.annotations.Nullable;
import org.openrewrite.Cursor;
import org.openrewrite.Tree;
import org.openrewrite.TreeVisitor;
import org.openrewrite.csharp.tree.CsDocComment;
import org.openrewrite.internal.ListUtils;
import org.openrewrite.java.tree.J;

public class CsDocCommentVisitor<P> extends TreeVisitor<CsDocComment, P> {
    /**
     * Don't use directly. Use {@link #csharpVisitorVisit(Tree, Object)} instead.
     */
    private final CSharpVisitor<P> csharpVisitor;

    public CsDocCommentVisitor(CSharpVisitor<P> csharpVisitor) {
        this.csharpVisitor = csharpVisitor;
    }

    protected @Nullable J csharpVisitorVisit(@Nullable Tree tree, P p) {
        Cursor previous = csharpVisitor.getCursor();
        J j = csharpVisitor.visit(tree, p, getCursor());
        csharpVisitor.setCursor(previous);
        return j;
    }

    public CsDocComment visitDocComment(CsDocComment.DocComment docComment, P p) {
        CsDocComment.DocComment d = docComment;
        d = d.withMarkers(visitMarkers(d.getMarkers(), p));
        return d.withBody(ListUtils.map(d.getBody(), b -> visit(b, p)));
    }

    public CsDocComment visitXmlElement(CsDocComment.XmlElement element, P p) {
        CsDocComment.XmlElement e = element;
        e = e.withMarkers(visitMarkers(e.getMarkers(), p));
        e = e.withAttributes(ListUtils.map(e.getAttributes(), attr -> visit(attr, p)));
        e = e.withSpaceBeforeClose(ListUtils.map(e.getSpaceBeforeClose(), s -> visit(s, p)));
        e = e.withContent(ListUtils.map(e.getContent(), c -> visit(c, p)));
        return e.withClosingTagSpaceBeforeClose(ListUtils.map(e.getClosingTagSpaceBeforeClose(), s -> visit(s, p)));
    }

    public CsDocComment visitXmlEmptyElement(CsDocComment.XmlEmptyElement element, P p) {
        CsDocComment.XmlEmptyElement e = element;
        e = e.withMarkers(visitMarkers(e.getMarkers(), p));
        e = e.withAttributes(ListUtils.map(e.getAttributes(), attr -> visit(attr, p)));
        return e.withSpaceBeforeSlashClose(ListUtils.map(e.getSpaceBeforeSlashClose(), s -> visit(s, p)));
    }

    public CsDocComment visitXmlText(CsDocComment.XmlText text, P p) {
        CsDocComment.XmlText t = text;
        return t.withMarkers(visitMarkers(t.getMarkers(), p));
    }

    public CsDocComment visitXmlAttribute(CsDocComment.XmlAttribute attribute, P p) {
        CsDocComment.XmlAttribute a = attribute;
        a = a.withMarkers(visitMarkers(a.getMarkers(), p));
        a = a.withSpaceBeforeEquals(ListUtils.map(a.getSpaceBeforeEquals(), s -> visit(s, p)));
        return a.withValue(ListUtils.map(a.getValue(), v -> visit(v, p)));
    }

    public CsDocComment visitXmlCrefAttribute(CsDocComment.XmlCrefAttribute attribute, P p) {
        CsDocComment.XmlCrefAttribute a = attribute;
        a = a.withMarkers(visitMarkers(a.getMarkers(), p));
        a = a.withSpaceBeforeEquals(ListUtils.map(a.getSpaceBeforeEquals(), s -> visit(s, p)));
        a = a.withValue(ListUtils.map(a.getValue(), v -> visit(v, p)));
        return a.withReference(csharpVisitorVisit(a.getReference(), p));
    }

    public CsDocComment visitXmlNameAttribute(CsDocComment.XmlNameAttribute attribute, P p) {
        CsDocComment.XmlNameAttribute a = attribute;
        a = a.withMarkers(visitMarkers(a.getMarkers(), p));
        a = a.withSpaceBeforeEquals(ListUtils.map(a.getSpaceBeforeEquals(), s -> visit(s, p)));
        a = a.withValue(ListUtils.map(a.getValue(), v -> visit(v, p)));
        return a.withParamName(csharpVisitorVisit(a.getParamName(), p));
    }

    public CsDocComment visitLineBreak(CsDocComment.LineBreak lineBreak, P p) {
        CsDocComment.LineBreak l = lineBreak;
        return l.withMarkers(visitMarkers(l.getMarkers(), p));
    }
}

File: rewrite-csharp/csharp/OpenRewrite/CSharp/CsDocCommentVisitor.cs
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using OpenRewrite.Core;
using OpenRewrite.Java;

namespace OpenRewrite.CSharp;

/// <summary>
/// Visitor over the structured C# XML documentation comment tree, mirroring the Java
/// <c>CsDocCommentVisitor</c>. Bridges to a <see cref="CSharpVisitor{P}"/> so <c>cref</c>/<c>name</c>
/// attribute references (which are ordinary <see cref="J"/> nodes) can be visited.
/// </summary>
public class CsDocCommentVisitor<P>
{
    private readonly CSharpVisitor<P> _csharpVisitor;

    public CsDocCommentVisitor(CSharpVisitor<P> csharpVisitor)
    {
        _csharpVisitor = csharpVisitor;
    }

    public Cursor Cursor { get; set; } = new();

    protected J? CsharpVisitorVisit(Tree? tree, P p)
    {
        if (tree == null) return null;
        var previous = _csharpVisitor.Cursor;
        var j = _csharpVisitor.Visit(tree, p, Cursor);
        _csharpVisitor.Cursor = previous;
        return j;
    }

    protected virtual Markers VisitMarkers(Markers markers, P p) => _csharpVisitor.VisitMarkers(markers, p);

    public virtual CsDocComment? Visit(CsDocComment? node, P p) => node switch
    {
        CsDocComment.DocComment d => VisitDocComment(d, p),
        CsDocComment.XmlElement e => VisitXmlElement(e, p),
        CsDocComment.XmlEmptyElement e => VisitXmlEmptyElement(e, p),
        CsDocComment.XmlText t => VisitXmlText(t, p),
        CsDocComment.XmlCrefAttribute a => VisitXmlCrefAttribute(a, p),
        CsDocComment.XmlNameAttribute a => VisitXmlNameAttribute(a, p),
        CsDocComment.XmlAttribute a => VisitXmlAttribute(a, p),
        CsDocComment.LineBreak l => VisitLineBreak(l, p),
        _ => node
    };

    protected IList<CsDocComment> VisitList(IList<CsDocComment> nodes, P p)
    {
        // Identity-preserving, mirroring Java's ListUtils.map: return the SAME list instance when
        // no element changed, so With*/change-detection upstream can short-circuit unchanged trees.
        List<CsDocComment>? result = null;
        for (int i = 0; i < nodes.Count; i++)
        {
            var original = nodes[i];
            var visited = Visit(original, p);
            if (result == null && !ReferenceEquals(visited, original))
            {
                result = new List<CsDocComment>(nodes.Count);
                for (int j = 0; j < i; j++) result.Add(nodes[j]);
            }
            if (result != null && visited != null) result.Add(visited);
        }
        return result ?? nodes;
    }

    protected IList<CsDocComment>? VisitListNullable(IList<CsDocComment>? nodes, P p) =>
        nodes == null ? null : VisitList(nodes, p);

    public virtual CsDocComment VisitDocComment(CsDocComment.DocComment docComment, P p) =>
        docComment.WithMarkers(VisitMarkers(docComment.Markers, p))
            .WithBody(VisitList(docComment.Body, p));

    public virtual CsDocComment VisitXmlElement(CsDocComment.XmlElement element, P p) =>
        element.WithMarkers(VisitMarkers(element.Markers, p))
            .WithAttributes(VisitList(element.Attributes, p))
            .WithSpaceBeforeClose(VisitList(element.SpaceBeforeClose, p))
            .WithContent(VisitList(element.Content, p))
            .WithClosingTagSpaceBeforeClose(VisitList(element.ClosingTagSpaceBeforeClose, p));

    public virtual CsDocComment VisitXmlEmptyElement(CsDocComment.XmlEmptyElement element, P p) =>
        element.WithMarkers(VisitMarkers(element.Markers, p))
            .WithAttributes(VisitList(element.Attributes, p))
            .WithSpaceBeforeSlashClose(VisitList(element.SpaceBeforeSlashClose, p));

    public virtual CsDocComment VisitXmlText(CsDocComment.XmlText text, P p) =>
        text.WithMarkers(VisitMarkers(text.Markers, p));

    public virtual CsDocComment VisitXmlAttribute(CsDocComment.XmlAttribute attribute, P p) =>
        attribute.WithMarkers(VisitMarkers(attribute.Markers, p))
            .WithSpaceBeforeEquals(VisitListNullable(attribute.SpaceBeforeEquals, p))
            .WithValue(VisitListNullable(attribute.Value, p));

    public virtual CsDocComment VisitXmlCrefAttribute(CsDocComment.XmlCrefAttribute attribute, P p) =>
        attribute.WithMarkers(VisitMarkers(attribute.Markers, p))
            .WithSpaceBeforeEquals(VisitListNullable(attribute.SpaceBeforeEquals, p))
            .WithValue(VisitListNullable(attribute.Value, p))
            .WithReference(CsharpVisitorVisit(attribute.Reference, p));

    public virtual CsDocComment VisitXmlNameAttribute(CsDocComment.XmlNameAttribute attribute, P p) =>
        attribute.WithMarkers(VisitMarkers(attribute.Markers, p))
            .WithSpaceBeforeEquals(VisitListNullable(attribute.SpaceBeforeEquals, p))
            .WithValue(VisitListNullable(attribute.Value, p))
            .WithParamName(CsharpVisitorVisit(attribute.ParamName, p));

    public virtual CsDocComment VisitLineBreak(CsDocComment.LineBreak lineBreak, P p) =>
        lineBreak.WithMarkers(VisitMarkers(lineBreak.Markers, p));
}
